### PR TITLE
enhancement(config): Implicit namespacing based on the config directory structure

### PR DESCRIPTION
**Pull request by jdrouet**
Signed-off-by: Jérémie Drouet <jeremie.drouet@gmail.com>

- [x] Update the loading process to load components to implement the [strategy design pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/strategy.html).
- [x] Load `transforms` from subfolder
- [x] Load `sinks` from subfolder
- [x] Load `sources` from subfolder
- [x] Load `enrichment_tables`  from subfolder
- [x] Load `tests` from subfolder

### CODE REVIEW

File: config/examples/namespacing/sinks/es_cluster.toml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Send structured data to a short-term storage
+inputs       = ["apache_sample"]            # only take sampled data
+type         = "elasticsearch"
+host         = "http://79.12.221.222:9200"   # local or external host
+index        = "vector-%Y-%m-%d"             # daily indices
+
+
```
Comment (jszwedko) [inline config/examples/namespacing/sinks/es_cluster.toml:7]: The smallest of nits: some of these files have trailing whitespace.

File: website/content/en/docs/reference/configuration/_index.md [hunk 2]
```diff
@@ -266,6 +266,94 @@ Or using a [globbing syntax][glob]:
 vector --config /etc/vector/*.toml
 ```
 
+#### Automatic namespacing
+
+You can also split your configuration by grouping the components by their type.
+The example can then be split as follow.
```
Comment (jszwedko) [inline website/content/en/docs/reference/configuration/_index.md:272]:
```suggestion
You can also split your configuration by grouping the components by their type, one directory per component type, where the file name is used as the component id. For example:
```

File: website/cue/reference/configuration.cue [hunk 3]
```diff
@@ -273,6 +273,28 @@ configuration: {
 				```
 				"""
 		}
+		automatic_namespacing: {
+			title: "Automatic namespacing of component files"
+			body: """
+				You can split your configuration files in component-type related folders.
+				That way, you can create the sink `foo` in the folder `/path/to/vector/config/sinks/foo.toml` and
+				configure it as follow.
```
Comment (jszwedko) [inline website/cue/reference/configuration.cue:281]:
```suggestion
				You can split your configuration files in component-type related folders.
				
				For example, you can create the sink `foo` in the folder `/path/to/vector/config/sinks/foo.toml` and
				configure it as follows:
```

File: website/content/en/guides/level-up/managing-complex-configs.md [hunk 4]
```diff
@@ -179,6 +179,70 @@ vector -c ./configs/foo.toml ./configs/bar.toml
 If you have a large chain of components it's a good idea to break them out into
 individual files, each with its own unit tests.
 
+## Splitting Configs
+
+If your components start to by used in multiple configuration files, having a
+dedicated place to define them can become interesting.
+
+With Vector you can define a component configuration inside a component type folder.
+Let's take an example with the following configuration file.
```
Comment (jszwedko) [inline website/content/en/guides/level-up/managing-complex-configs.md:188]:
```suggestion
With Vector you can define a component configuration inside a component type folder.

Let's take an example with the following configuration file:
```

**Review by jszwedko — COMMENTED** [review]
I left a few comments about the docs, but otherwise this looks good!

File: website/content/en/guides/level-up/managing-complex-configs.md [hunk 5]
```diff
@@ -179,6 +179,71 @@ vector -c ./configs/foo.toml ./configs/bar.toml
...
+And the `stdout` sink in the file `/etc/vector/sinks/stdout.toml`
+
+```toml title="stdout.toml"
+type = "console"
+inputs = ["change_fields"]
+target = "stdout"
+```
+
+And for Vector to look for the configuration in the component type related folders,
+you need to start is using the `--config-dir` argument as follow.
```
Comment (vladimir-dd) [inline website/content/en/guides/level-up/managing-complex-configs.md:241]:
```suggestion
you need to start it using the `--config-dir` argument as follows.
```

File: src/config/loading.rs [hunk 6]
```diff
@@ -251,3 +398,28 @@ pub fn load(
...
     format::deserialize(&with_vars, format).map(|builder| (builder, warnings))
 }
+
+#[cfg(test)]
+mod tests {
+    use super::load_builder_from_paths;
+    use crate::config::{ComponentKey, ConfigPath};
+    use std::path::PathBuf;
+
+    #[test]
```
Comment (vladimir-dd) [inline src/config/loading.rs:408]: [nit] Can we also test a negative scenario here - e.g. an unknown component key?

**Review by vladimir-dd — APPROVED** [review]
Neat! Looks good to me.

**Review by jszwedko — APPROVED** [review]
Nice!

**Review by binarylogic — APPROVED** [review]
Looks great! Thanks for the docs changes.
